File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportJava/Persistence/src/main/java/transport/persistence/repository/jdbc/UserDbRepository.java
package transport.persistence.repository.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import transport.model.User;
import transport.persistence.repository.UserRepository;
import transport.persistence.repository.utils.JdbcUtils;
import transport.persistence.repository.utils.PasswordUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class UserDbRepository implements UserRepository {
    private JdbcUtils dbUtils;

    private static final Logger logger = LogManager.getLogger();

    public UserDbRepository(Properties properties) {
        logger.info("Initializind UserDbRepository with properties: {} ", properties);
        dbUtils = new JdbcUtils(properties);
    }

    @Override
    public User findByUsername(String username) {
        logger.info("Attempting login for user: {}", username);
        String sql = "SELECT * FROM users WHERE username = ? LIMIT 1    ";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, username);
            try(var result = statement.executeQuery()) {
                if(result.next()) {
                    String password = result.getString("password");
                    User user = new User(result.getLong("id"), username, password);
                    logger.trace("User found: {}", user);
                    return user;
                } else {
                    logger.trace("No user found with username: {}", username);
                    return null;
                }
            }
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error during finding by username: " + e);
            return null;
        }
    }

    @Override
    public Long add(User entity) {
        logger.info("Adding new user: {}", entity);
        String sql = "INSERT INTO users (username, password) VALUES (?, ?)";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {
            statement.setString(1, entity.getUsername());
            statement.setString(2, PasswordUtils.hash(entity.getPassword()));
            int result = statement.executeUpdate();
            logger.trace("User added successfully, result: {}", result);

            try (ResultSet generatedKeys = statement.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    Long id = generatedKeys.getLong(1);
                    entity.setId(id);
                    return id;
                }
            }
            return null;
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error adding user: " + e);
            return null;
        } finally {
            logger.traceExit();
        }
    }

    @Override
    public void delete(Long aLong) {
        logger.info("Deleting user: {}", aLong);
        String sql = "DELETE FROM users WHERE id = ?";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, aLong);
            int result = statement.executeUpdate();
            logger.trace("User deleted successfully, result: {}", result);
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error deleting user: " + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(User entity) {
        logger.info("Updating user: {}", entity);
        String sql = "UPDATE users SET username = ?, password = ? WHERE id = ?";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, entity.getUsername());
            statement.setString(2, PasswordUtils.hash(entity.getPassword()));
            statement.setLong(3, entity.getId());
            int result = statement.executeUpdate();
            logger.trace("User updated successfully, result: {}", result);
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error updating user: " + e);
        }
        logger.traceExit();
    }

    @Override
    public User findById(Long aLong) {
        logger.info("Finding user: {}", aLong);
        String sql = "SELECT * FROM users WHERE id = ?";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, aLong);
            try(var result = statement.executeQuery()) {
                if (result.next()) {
                    String username = result.getString("username");
                    String password = result.getString("password");
                    User user = new User(aLong, username, password);
                    logger.trace("User found: {}", user);
                    return user;
                } else {
                    logger.trace("User not found with id: {}", aLong);
                    return null;
                }
            }
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error finding user: " + e);
            return null;
        }
    }

    @Override
    public List<User> findAll() {
        logger.info("Finding all users");
        String sql = "SELECT * FROM users";
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement statement = connection.prepareStatement(sql)) {
            try(var result = statement.executeQuery()) {
                List<User> users = new ArrayList<>();
                while (result.next()) {
                    Long id = result.getLong("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    User user = new User(id, username, password);
                    users.add(user);
                    logger.trace("User found: {}", user);
                }
                return users;
            }
        } catch (Exception e) {
            logger.error(e);
            System.out.println("Error finding users: " + e);
        }
        logger.traceExit();
        return new ArrayList<>();
    }
}

File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportC#/Persistence/ado/UserDbRepository.cs
using FirmaTransportC_.Model;
using FirmaTransportC_.Persistence.utils;
using log4net;
using System.Data;

namespace FirmaTransportC_.Persistence.ado
{
    public class UserDbRepository : IUserRepository
    {
        private static readonly ILog log = LogManager.GetLogger("UserDbRepository");

        IDictionary<String, string> props;

        public UserDbRepository(IDictionary<String, string> props)
        {
            log.Info("Creating UserDbRepository with properties: " + string.Join(", ", props));
            this.props = props;
        }

        public long Add(User entity)
        {
            log.InfoFormat("Adding new user: {0}", entity);

            using (IDbConnection connection = DbUtils.GetConnection(props))
            {
                try
                {
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = "INSERT INTO users (username, password) VALUES (@username, @password); SELECT last_insert_rowid();";

                        var paramUsername = command.CreateParameter();
                        paramUsername.ParameterName = "@username";
                        paramUsername.Value = entity.Username;
                        command.Parameters.Add(paramUsername);

                        var paramPassword = command.CreateParameter();
                        paramPassword.ParameterName = "@password";
                        paramPassword.Value = PasswordUtils.Hash(entity.Password);
                        command.Parameters.Add(paramPassword);

                        var result = command.ExecuteScalar();

                        if (result != null)
                        {
                            long id = Convert.ToInt64(result);
                            entity.Id = id;

                            log.InfoFormat("User added successfully, id: {0}", id);
                            return id;
                        }
                    }
                }
                catch (Exception ex)
                {
                    log.Error("Error adding user", ex);
                }
            }

            return -1;
        }

        public void Delete(long id)
        {
            log.InfoFormat("Entering Delete with id: {0}", id);
            IDbConnection connection = DbUtils.GetConnection(props);

            try
            {
                int result;
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "DELETE FROM users WHERE id = @id";
                    var paramId = command.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    command.Parameters.Add(paramId);

                    result = command.ExecuteNonQuery();
                }
                log.InfoFormat("Exiting Delete with result: {0}", result);
            }
            catch (Exception ex)
            {
                log.Error("Error in Delete", ex);
            }
            finally
            {
                connection.Close();
            }
        }

        public List<User> FindAll()
        {
            log.Info("Entering FindAll");
            IDbConnection connection = DbUtils.GetConnection(props);
            var users = new List<User>();

            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM users";
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            users.Add(new User(
                                reader.GetInt64(0),
                                reader.GetString(1),
                                reader.GetString(2)
                            ));
                        }
                    }
                }
                log.InfoFormat("Exiting FindAll with {0} users", users.Count);
            }
            catch (Exception ex)
            {
                log.Error("Error in FindAll", ex);
            }
            finally
            {
                connection.Close();
            }

            return users;
        }

        public User FindOne(long id)
        {
            log.InfoFormat("Entering FindOne with id: {0}", id);
            IDbConnection connection = DbUtils.GetConnection(props);

            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM users WHERE id = @id";
                    var paramId = command.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    command.Parameters.Add(paramId);
                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            var user = new User(
                                reader.GetInt64(0),
                                reader.GetString(1),
                                reader.GetString(2)
                            );
                            log.InfoFormat("Exiting FindOne with result {0}", user);
                            return user;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error in FindOne", ex);
            }
            finally
            {
                connection.Close();
            }
            return null;
        }

        public User? FindByUsername(string username)
        {
            log.InfoFormat("Entering FindByUsername with username: {0}", username);
            IDbConnection connection = DbUtils.GetConnection(props);

            try
            {
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "SELECT * FROM users WHERE username = @username LIMIT 1";

                    var paramUsername = command.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = username;
                    command.Parameters.Add(paramUsername);

                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            var user = new User(
                                reader.GetInt64(0),
                                reader.GetString(1),
                                reader.GetString(2)
                            );
                            log.InfoFormat("Exiting FindByUsername with result: {0}", user);
                            return user;
                        }
                        return null;
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Error in Login", ex);
                return null;
            }
            finally
            {
                connection.Close();
            }
        }

        public void Update(User entity)
        {
            log.InfoFormat("Entering Update with user: {0}", entity);
            IDbConnection connection = DbUtils.GetConnection(props);

            try
            {
                int result;
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = "UPDATE users SET username = @username, password = @password WHERE id = @id";

                    var paramUsername = command.CreateParameter();
                    paramUsername.ParameterName = "@username";
                    paramUsername.Value = entity.Username;
                    command.Parameters.Add(paramUsername);

                    var paramPassword = command.CreateParameter();
                    paramPassword.ParameterName = "@password";
                    paramPassword.Value = entity.Password;
                    command.Parameters.Add(paramPassword);

                    var paramId = command.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = entity.Id;
                    command.Parameters.Add(paramId);

                    result = command.ExecuteNonQuery();
                }
                log.InfoFormat("Exiting Update with result: {0}", result);
            }
            catch (Exception ex)
            {
                log.Error("Error in Update", ex);
            }
            finally
            {
                connection.Close();
            }
        }
    }
}
